File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/EnumMember.cs
using System.Collections.Generic;
using Lexy.Poc.Core.Parser;
using Lexy.Poc.Core.Parser.Tokens;
using Microsoft.CodeAnalysis.CSharp;

namespace Lexy.Poc.Core.Language
{
    public class EnumMember : Node
    {
        public NumberLiteralToken Value { get; }
        public string Name { get; }

        public EnumMember(string name, NumberLiteralToken value = null)
        {
            Value = value;
            Name = name;
        }

        public static EnumMember Parse(IParserContext context)
        {
            var valid = context.ValidateTokens<AssignmentDefinition>()
                .CountMinimum(1)
                .StringLiteral(0)
                .IsValid;

            if (!valid) return null;

            var tokens = context.CurrentLine.Tokens;
            var name = tokens.TokenValue(0);
            if (tokens.Length == 1)
            {
                return new EnumMember(name);
            }

            if (tokens.Length != 3)
            {
                context.Logger.Fail(
                    $"Invalid number of tokens: {tokens.Length}. Should be 1 or 3.");
                return null;
            }

            valid = context.ValidateTokens<AssignmentDefinition>()
                .Operator(1, OperatorType.Assignment)
                .NumberLiteral(2)
                .IsValid;
            if (!valid) return null;

            var value = tokens.Token<NumberLiteralToken>(2);
            return new EnumMember(name, value);
        }

        protected override IEnumerable<INode> GetChildren()
        {
            yield break;
        }

        protected override void Validate(IParserContext context)
        {
            ValidateMemberName(context);
            ValidateMemberValues(context);
        }

        private void ValidateMemberName(IParserContext context)
        {
            if (string.IsNullOrEmpty(Name))
            {
                context.Logger.Fail("Enum member name should not be null or empty.");
            }
            else if (!SyntaxFacts.IsValidIdentifier(Name))
            {
                context.Logger.Fail($"Invalid enum member name: {Name}.");
            }
        }

        private void ValidateMemberValues(IParserContext context)
        {
            if (Value == null) return;

            if (Value.NumberValue < 0)
            {
                context.Logger.Fail("Enum member value should not be < 0: " + Value);
            }

            if (Value.IsDecimal())
            {
                context.Logger.Fail("Enum member value should not be decimal: " + Value);
            }
        }
    }
}

File: typescipt/existing/src/language/enums/enumMember.ts





namespace Lexy.Compiler.Language.Enums;

public class EnumMember : Node
{
   public string Name { get; }
   public NumberLiteralToken ValueLiteral { get; }
   public int NumberValue { get; }

   private EnumMember(string name, SourceReference reference, NumberLiteralToken valueLiteral, int value) :
     base(reference)
   {
     NumberValue = value;
     Name = name;
     ValueLiteral = valueLiteral;
   }

   public static EnumMember Parse(IParseLineContext context, int lastIndex)
   {
     var valid = context.ValidateTokens<EnumMember>()
       .CountMinimum(1)
       .StringLiteral(0)
       .IsValid;

     if (!valid) return null;

     var line = context.Line;
     var tokens = line.Tokens;
     var name = tokens.TokenValue(0);
     var reference = line.LineStartReference();

     if (tokens.Length = 1) return new EnumMember(name, reference, null, lastIndex + 1);

     if (tokens.Length ! 3)
     {
       context.Logger.Fail(line.LineEndReference(),
         $"Invalid number of tokens: {tokens.Length}. Should be 1 or 3");
       return null;
     }

     valid = context.ValidateTokens<EnumMember>()
       .Operator(1, OperatorType.Assignment)
       .NumberLiteral(2)
       .IsValid;
     if (!valid) return null;

     var value = tokens.Token<NumberLiteralToken>(2);

     return new EnumMember(name, reference, value, (int)value.NumberValue);
   }

   public override IEnumerable<INode> GetChildren()
   {
     yield break;
   }

   protected override void Validate(IValidationContext context)
   {
     ValidateMemberName(context);
     ValidateMemberValues(context);
   }

   private void ValidateMemberName(IValidationContext context)
   {
     if (string.IsNullOrEmpty(Name))
       context.Logger.Fail(Reference, "Enum member name should not be null or empty.");
     else if (!SyntaxFacts.IsValidIdentifier(Name))
       context.Logger.Fail(Reference, $"Invalid enum member name: {Name}.");
   }

   private void ValidateMemberValues(IValidationContext context)
   {
     if (ValueLiteral = null) return;

     if (ValueLiteral.NumberValue < 0)
       context.Logger.Fail(Reference, $"Enum member value should not be < 0: {ValueLiteral}");

     if (ValueLiteral.IsDecimal())
       context.Logger.Fail(Reference, $"Enum member value should not be decimal: {ValueLiteral}");
   }
}
